File: src/install/prompt.mjs
// prompt.mjs — the minimal interactive selection UI for `landfall install`
// (specs/049-cli-mcp-harness-installer/research.md R4). A small readline-based
// per-item confirm, not a full raw-mode checkbox widget: it's simple to test
// (feed it lines) and matches this library's near-zero dependency convention.
//
// Deliberately consumes `input` as an async iterator (one line per question)
// rather than repeated `readline.question()` calls: `question()` from
// `node:readline/promises` is unreliable once the underlying stream reaches
// EOF (exactly what a piped, non-interactive `stdin` does immediately after
// all its lines are written) — a later `question()` call can hang forever
// instead of resolving with the already-buffered line. Iterating the stream
// directly has no such failure mode and works identically for a real TTY.
import readline from 'node:readline';

/**
 * Ask the user, one at a time, whether to select each item (all pre-checked
 * — pressing Enter accepts the default). Returns the array of selected item
 * ids, preserving `items`' order.
 *
 * @param {{id: string, label: string}[]} items
 * @param {{skipPrompt?: boolean, input?: NodeJS.ReadableStream, output?: NodeJS.WritableStream}} [opts]
 *   `skipPrompt: true` (the `--yes` flag) selects every item with no I/O at all.
 */
export async function promptSelection(items, { skipPrompt = false, input = process.stdin, output = process.stdout } = {}) {
  if (skipPrompt || items.length === 0) return items.map((i) => i.id);

  const rl = readline.createInterface({ input });
  const lines = rl[Symbol.asyncIterator]();
  const selected = [];
  try {
    for (const item of items) {
      output.write(`Configure ${item.label}? [Y/n] `);
      const { value, done } = await lines.next();
      const answer = done ? '' : String(value).trim().toLowerCase();
      if (!answer.startsWith('n')) selected.push(item.id);
    }
  } finally {
    rl.close();
  }
  return selected;
}

File: internal/install/prompt.go
// prompt.go — the minimal interactive selection UI for `landfall install`.
// A Go port of `src/install/prompt.mjs`.
//
// A small line-at-a-time per-item confirm, not a full raw-mode checkbox
// widget: it is simple to test (feed it lines) and matches this CLI's
// near-zero-dependency convention.
//
// The Node original goes out of its way to consume stdin as an async ITERATOR
// rather than through repeated `readline.question()` calls, because
// `question()` is unreliable once the underlying stream hits EOF — exactly
// what a piped, non-interactive stdin does the moment its lines are written,
// and a later call could hang forever instead of resolving with an
// already-buffered line. Go has no such hazard: a bufio.Scanner over the
// stream reads buffered lines and then reports EOF, which this treats as the
// same "" answer the Node version's `done` case produces. The behaviour the
// comment protects — every remaining item still gets its default answer after
// input runs out, and nothing blocks — is preserved.
package install

import (
	"bufio"
	"fmt"
	"io"
	"strings"
)

// SelectItem is one selectable harness/host.
type SelectItem struct {
	ID    string
	Label string
}

// PromptSelection asks the user, one at a time, whether to select each item.
// All are pre-checked: pressing Enter accepts the default, and only an answer
// starting with "n" declines. Returns the selected ids, preserving `items`
// order.
//
// skipPrompt (the `--yes` flag) selects every item with no I/O AT ALL — not
// even a read — which is what lets `--yes` run in a script with no stdin
// attached without hanging.
func PromptSelection(items []SelectItem, skipPrompt bool, in io.Reader, out io.Writer) []string {
	ids := make([]string, 0, len(items))
	if skipPrompt || len(items) == 0 {
		for _, i := range items {
			ids = append(ids, i.ID)
		}
		return ids
	}

	scanner := bufio.NewScanner(in)
	for _, item := range items {
		fmt.Fprintf(out, "Configure %s? [Y/n] ", item.Label)
		answer := ""
		if scanner.Scan() {
			answer = strings.ToLower(strings.TrimSpace(scanner.Text()))
		}
		if !strings.HasPrefix(answer, "n") {
			ids = append(ids, item.ID)
		}
	}
	return ids
}
